language/tools/read-write-set/types/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+mod access;
+
+pub use access::Access;
+
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::StructType,
+};
+use std::collections::{hash_map::Entry, HashMap};
+
+/// Offset of an access path: either a field, vector index, or global key
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum Offset {
+    /// Index into contents of a struct by field offset
+    Field(usize),
+    /// Unknown index into a vector
+    VectorIndex,
+    /// A type index into global storage. Only follows a field or vector index of type address
+    Global(StructType),
+}
+
+#[derive(Debug, Clone)]
+pub struct TrieNode {
+    /// Optional data associated with the parent in the trie
+    data: Option<Access>,
+    /// Child pointers labeled by offsets
+    children: HashMap<Offset, TrieNode>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Root {
+    Const(AccountAddress),
+    Formal(usize),
+}
+
+#[derive(Debug, Clone)]
+pub struct AccessPath {
+    pub root: Root,
+    pub offsets: Vec<Offset>,
+}
+
+pub struct ReadWriteSet(HashMap<Root, TrieNode>);
```

Comment:
This is the core definition for the inferred read-writeset result. It's a bit unfortunate that we can't reuse the code from `move-model` due to all the references to `GlobalEnv`.